napari/utils/events/evented_model.py

```diff
@@ -181,16 +183,21 @@ class Config:
         # False, RuntimeError will be raised on model declaration
         arbitrary_types_allowed = True
         # whether to perform validation on assignment to attributes
+        # note that this should stay True for Field(allow_mutation) to work.
         validate_assignment = True
         # whether to treat any underscore non-class var attrs as private
         # https://pydantic-docs.helpmanual.io/usage/models/#private-model-attributes
         underscore_attrs_are_private = True
         # whether to validate field defaults (default: False)
+        # see https://github.com/napari/napari/pull/4138 before changing.
         validate_all = True
         # https://pydantic-docs.helpmanual.io/usage/exporting_models/#modeljson
         # NOTE: json_encoders are also added EventedMetaclass.__new__ if the
         # field declares a _json_encode method.
         json_encoders = _BASE_JSON_ENCODERS
+        # this custom use of allow mutation behaves as normal (0 is false, 1 and 2 are true)
+        # but allows us to use 2 as a special value meaning "inplace mutation"
+        allow_mutation = 2
```

Comment:
0, 1, and 2 will never be easy to read in code.  unless they are _always_ accompanied by an explanation of what each integer means (which you have found yourself needing to do in multiple places in this PR...), new readers will presumably always have to stop reading and go search the code for what they mean.  Consider an enum with a custom `__bool__` method, or a string and a convenience `is_true()` function

Reply:
Add named contraints in pydantic_extra and import it in every place that is required

Reply:
Inplace mutation is not something that is part of what `extra-pydantic` does, it's all napari (necessary due to our evented objects). So this should happen here.